bayem/vba.py

```diff
@@ -0,0 +1,519 @@
+import logging
+from dataclasses import dataclass
+from time import perf_counter
+from tabulate import tabulate
+from typing import Union, Dict, Tuple
+
+import numpy as np
+import scipy.special as special
+
+from .distributions import MVN, Gamma
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass
+class Options:
+    tolerance: float = 0.1
+    maxiter: int = 50
+    maxtrials: int = 10
+    update_noise: Union[Dict, bool] = True
+    index_ARD: Tuple[int] = ()
+
+
+class CDF_Jacobian:
+    def __init__(self, f, transformation):
+        self._f = f
+        self._T = transformation
+
+    def __call__(self, _x):
+        x = np.copy(_x)
+
+        for iParam in range(len(x)):
+            eps = 1.0e-7  # approx sqrt(machine precision)
+            dx = max(eps, abs(x[iParam]) * eps)
+
+            x[iParam] -= dx
+            fs0 = self._T(self._f(x))
+            x[iParam] += 2 * dx
+            fs1 = self._T(self._f(x))
+            x[iParam] = _x[iParam]
+
+            if iParam == 0:
+                # allocate jac
+                jac = {}
+                for key, f0 in fs0.items():
+                    jac[key] = np.empty([len(f0), len(x)])
+
+            for n in fs0:
+                jac[n][:, iParam] = (fs1[n] - fs0[n]) / (2 * dx)
+
+        return jac
+
+
+class VBAProblem:
+    def __init__(self, f, noise0, jac):
+        self.f = f
+        self.noise0 = noise0
+        self.jac = jac
+        self.jac_in_f = not callable(jac) and jac
```

Comment:
isn't the not wrong? jac should be provided and callable=True?

Reply:
`jac_in_f` means that the call to `f` already contains the Jacobian, e.g.

~~~py
def my_model_error(theta):
   x = theta[0]
   return (time_steps * x**2, time_steps * 2*x)
~~~

In that case, the method is called with
~~~py
bayem.vba(my_model_error, x0, jac=True)
~~~